internal/internal_task_pollers.go

```diff
@@ -425,103 +431,114 @@ func (wtp *workflowTaskPoller) RespondTaskCompletedWithMetrics(completedRequest
 	return
 }
 
-func (wtp *workflowTaskPoller) RespondTaskCompleted(completedRequest interface{}, task *s.PollForDecisionTaskResponse) (response *s.RespondDecisionTaskCompletedResponse, err error) {
+func (wtp *workflowTaskPoller) respondTaskCompleted(completedRequest interface{}, task *s.PollForDecisionTaskResponse) (response *s.RespondDecisionTaskCompletedResponse, err error) {
 	ctx := context.Background()
 	// Respond task completion.
 	err = backoff.Retry(ctx,
 		func() error {
-			tchCtx, cancel, opt := newChannelContext(ctx, wtp.featureFlags)
-			defer cancel()
-			var err1 error
-			switch request := completedRequest.(type) {
-			case *s.RespondDecisionTaskFailedRequest:
-				// Only fail decision on first attempt, subsequent failure on the same decision task will timeout.
-				// This is to avoid spin on the failed decision task. Checking Attempt not nil for older server.
-				if task.Attempt != nil && task.GetAttempt() == 0 {
-					err1 = wtp.service.RespondDecisionTaskFailed(tchCtx, request, opt...)
-					if err1 != nil {
-						traceLog(func() {
-							wtp.logger.Debug("RespondDecisionTaskFailed failed.", zap.Error(err1))
-						})
-					}
+			response, err = wtp.respondTaskCompletedAttempt(completedRequest, task)
+			return err
+		}, createDynamicServiceRetryPolicy(ctx), isServiceTransientError)
+
+	return response, err
+}
+
+func (wtp *workflowTaskPoller) respondTaskCompletedAttempt(completedRequest interface{}, task *s.PollForDecisionTaskResponse) (*s.RespondDecisionTaskCompletedResponse, error) {
+	ctx, cancel, _ := newChannelContext(context.Background(), wtp.featureFlags)
```

Comment:
tbh I think I'd just save `opts` here and pass it through below, since it's just being thrown away and then duplicated 3x now.  it'll probably be more flexible for the future too, if we ever pass more args through, because it'll be 3 fewer places where they might get dropped.

Reply:
Opts are entirely independent from the context, and passing them around will recreate a slice 3+ times.

I'd vote for removing feature flags from this builder completely.

Reply:
I returned them. I kind of understand that it is possible for us to have feature flags that might go into the context. However, we don't, and this functionality looks completely orthogonal to the context right now.

Reply:
yea, at the moment I think it's separate enough that it's kinda strange that we pass them in like this at all.  so I think it's fine either way.

maybe yarpc has an "add these opts to every call" option somewhere nowadays.  we could do it with a middleware, but that's a pain too.